numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -2192,18 +2192,31 @@ array_fromfile(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
 static PyObject *
 array_fromiter(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
 {
-    PyObject *iter;
+    PyObject *iterable;
     Py_ssize_t nin = -1;
-    static char *kwlist[] = {"iter", "dtype", "count", NULL};
+    static char *kwlist[] = {"iterable", "dtype", "count", NULL};
+    static char *kwlist_old[] = {"iter", "dtype", "count", NULL};
     PyArray_Descr *descr = NULL;
+    PyObject *ptype, *pvalue, *ptraceback;
 
     if (!PyArg_ParseTupleAndKeywords(args, keywds,
                 "OO&|" NPY_SSIZE_T_PYFMT ":fromiter", kwlist,
-                &iter, PyArray_DescrConverter, &descr, &nin)) {
+                &iterable, PyArray_DescrConverter, &descr, &nin)) {
         Py_XDECREF(descr);
-        return NULL;
+
+        PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+        if (!PyArg_ParseTupleAndKeywords(args, keywds,
+                    "OO&|" NPY_SSIZE_T_PYFMT ":fromiter", kwlist_old,
+                    &iterable, PyArray_DescrConverter, &descr, &nin)) {
+            PyErr_Restore(ptype, pvalue, ptraceback);
+            return NULL;
+        }
+        if (DEPRECATE("The 'iter' keyword argument is deprecated. "
```

Comment:
Deprecations should have something like `/* NumPy 1.15, 2018-02-15 */` on the line before them - and preferably in the test too.